server/src/main/java/org/elasticsearch/action/admin/cluster/settings/TransportClusterUpdateSettingsAction.java

```diff
@@ -62,22 +68,59 @@ public TransportClusterUpdateSettingsAction(TransportService transportService, C
+     * Only at least one of cluster.blocks.read_only or cluster.blocks.read_only_allow_delete is being cleared (set to null or false).
+     * Or all of the following are true:
+     * 1. At least one of cluster.blocks.read_only or cluster.blocks.read_only_allow_delete is being cleared (set to null or false).
+     * 2. Neither cluster.blocks.read_only nor cluster.blocks.read_only_allow_delete is being set to true.
+     * 3. The only other settings in this update are archived ones being set to null.
+     */
     @Override
     protected ClusterBlockException checkBlock(ClusterUpdateSettingsRequest request, ClusterState state) {
-        // allow for dedicated changes to the metadata blocks, so we don't block those to allow to "re-enable" it
-        if (request.transientSettings().size() + request.persistentSettings().size() == 1) {
-            // only one setting
-            if (Metadata.SETTING_READ_ONLY_SETTING.exists(request.persistentSettings())
-                || Metadata.SETTING_READ_ONLY_SETTING.exists(request.transientSettings())
-                || Metadata.SETTING_READ_ONLY_ALLOW_DELETE_SETTING.exists(request.transientSettings())
-                || Metadata.SETTING_READ_ONLY_ALLOW_DELETE_SETTING.exists(request.persistentSettings())) {
-                // one of the settings above as the only setting in the request means - resetting the block!
+        Map<String, String> clearedBlockAndArchivedSettings = new HashMap<>();
+        if (checkClearedBlockAndArchivedSettings(request.transientSettings(), clearedBlockAndArchivedSettings)
+            && checkClearedBlockAndArchivedSettings(request.persistentSettings(), clearedBlockAndArchivedSettings)) {
+            if (clearedBlockAndArchivedSettings.containsKey(Metadata.SETTING_READ_ONLY_SETTING.getKey())
+                || clearedBlockAndArchivedSettings.containsKey(Metadata.SETTING_READ_ONLY_ALLOW_DELETE_SETTING.getKey())) {
                 return null;
             }
         }
+
         return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
     }
 
+    /**
+     * Check settings that only contains block and archived settings.
+     * @param settings target settings to be checked.
+     * @param clearedBlockAndArchivedSettings block settings that have been set to null or false,
+     *                                        archived settings that have been set to null.
+     * @return true if all settings are clear blocks or archived settings.
+     */
+    private boolean checkClearedBlockAndArchivedSettings(final Settings settings,
+                                                         final Map<String, String> clearedBlockAndArchivedSettings) {
+        for (String key : settings.keySet()) {
+            String value = settings.get(key);
+            if (Metadata.SETTING_READ_ONLY_SETTING.getKey().equals(key)
+                || Metadata.SETTING_READ_ONLY_ALLOW_DELETE_SETTING.getKey().equals(key)) {
+                if ("true".equals(value)) {
```

Comment:
I don't think we should duplicate the parsing of boolean settings here. Suggest using `Metadata.SETTING_READ_ONLY_ALLOW_DELETE_SETTING.get(settings)` etc.